Inject logger instance via constructor

tux-rampage opened the pull request:
PR #32 introduces logging capabilities by the use of an "Aware" interface (`LoggerAwareInterface` of PSR-3). This is discouraged in ZF.

This pull request changes it to constructor injection and removes the introduced usage of *Aware*.

Since this fixes a violation that only exists in the develop branch, this PR is targeted against develop instead of master.

- [x] Are you fixing a bug?
  - [x] Detail how the bug is invoked currently.
  - [x] Detail the original, incorrect behavior.
  - [x] Detail the new, expected behavior.
  - [x] Add a regression test that demonstrates the bug, and proves the fix.
  - [x] Add a `CHANGELOG.md` entry for the fix.

- [x] Is this related to quality assurance?
  As stated by @Ocramius in PR #32, the usage of aware-interfaces are discouraged throughout the Framework

tux-rampage commented on .gitignore:
```diff
@@ -5,3 +5,4 @@
 /vendor/
 /zf-mkdoc-theme.tgz
 /zf-mkdoc-theme/
+/.idea
```
Please don't mind this change, it will be reverted when I complete the PR

Ocramius replied:
:+1:

Ocramius commented on test/CodeGenerator/InjectorGeneratorTest.php:
```diff
@@ -74,11 +74,10 @@ public function testGeneratorLogsDebugForEachClass()
-        $generator = new InjectorGenerator($config, $resolver);
         $logger = $this->prophesize(LoggerInterface::class);
 
+        $generator = new InjectorGenerator($config, $resolver, null, $logger->reveal());
```
Did you also change the constructor, or was it already accepting an optional logger?

tux-rampage replied:
This was the test, only. Just commited the code changes ;-)

tux-rampage commented on CHANGELOG.md:
```diff
@@ -6,6 +6,10 @@ All notable changes to this project will be documented in this file, in reverse
+  `Psr\Log\LoggerInterface` to `Zend\Di\CodeGenerator\InjectorGenerator`'s constructor
+
+
 - [#32](https://github.com/zendframework/zend-di/pull/32) adds the implementation of
```
@Ocramius I'm a bit unsure about leaving this in the changelog. Should it be removed as for readability (Same goes for the Removed and Fixed section). Since the changes from PR #32 were never released.

Ocramius replied:
Note is fine as long as you aggregate it with the changes in this patch. What matters is that the reader of the release notes understands what happened.

tux-rampage replied:
I guess it would be more confusing then, since it may lead people to the assumption that 3.1 introduces the Aware implementation which is actually not the case. I moved the reference to PR #32 to the Removed and Fixed entries to state clearly that this is not the case.

tux-rampage commented on CHANGELOG.md:
```diff
@@ -6,6 +6,10 @@ All notable changes to this project will be documented in this file, in reverse
+- [#33](https://github.com/zendframework/zend-di/pull/32) add ability to pass
+  `Psr\Log\LoggerInterface` to `Zend\Di\CodeGenerator\InjectorGenerator`'s constructor
+
+
```
Noted: Double space willl be removed -> single space.

tux-rampage commented:
Just for the records:

I chose the Aware implementation of PSR-3 since it is aliving PSR standard and it states a purely optional dependency that does not affect the classes core behaviour. But aware implementations are discouraged. Since this was already discussed - it will be removed without further arguing. (@Ocramius a ref would be awesome, if you got it at hand).

@Ocramius just two things to prevent this in the future and reference potential PR's to this:

1. Can you please specify your disagreement with the following line any further in the following terms https://github.com/zendframework/zend-di/pull/32/files#diff-6f24fd8c91ea4350af062354fa9081d2R19
* Is third party use of traits discouraged*
* Because it implements aware?
* Anything else?
2. I assume a setter without an Aware implementation would be ok to make life easier for consumers that want to decorate an already existing instance?

Ocramius commented:
> Is third party

Minor issue, given the interface stability

 > use of traits discouraged

Always discouraged. Only ever use traits in tests, as any downstream users may also use them and therefore we increase the BC surface by A LOT.

 > Because it implements aware?

This is the bigger issue - there is no need to expose that, as it is an internal implementation detail. Also, dependencies should not be mutable, and the `*aware*` interfaces render internal components mutable.

Ocramius commented:
> I assume a setter without an Aware implementation would be ok

The existence of the  setter is actually the bigger issue here.

tux-rampage commented:
Thanks a lot for the clearification. :+1: 

>> use of traits discouraged
>
> Always discouraged. 

The internal use only, could be stated with an "@internal" annotation, but I got your point. Even though it was marked internal someone will use it and blame the framework when the trait changes. I'll open a PR removing remaining traits in this component (but this will be low prio). By now no more introduction of new traits.

> The existence of the setter is actually the bigger issue here.

Even though it is not an actual dependency, more a utility aggregation (the generator's functionality itself does not actually depend on a logger)? I find this a bit too dogmatic, but I'm fine with cutting this.

webimpress commented on src/CodeGenerator/InjectorGenerator.php:
```diff
@@ -60,25 +61,33 @@ class InjectorGenerator implements LoggerAwareInterface
      * @param DependencyResolverInterface $resolver The resolver to utilize
      * @param string $namespace Namespace to use for generated class; defaults
      *     to Zend\Di\Generated.
+     * @param LoggerInterface $logger An optional logger instance to log failures
```
Missing `null` type in php doc.

tux-rampage replied:
Correct, on $namespace as well ...

webimpress commented on CHANGELOG.md:
```diff
@@ -27,11 +27,14 @@ All notable changes to this project will be documented in this file, in reverse
-- Nothing.
+- [#33](https://github.com/zendframework/zend-di/pull/32) removes the implementation of
+  `Psr\Log\LoggerAwareInterface` from `Zend\Di\CodeGenerator\InjectorGenerator` that
+  was introduced with [#32](https://github.com/zendframework/zend-di/pull/32)
```
I think we should remove it, because #32 is not going to be released, because this PR replaces it. The same in `Fixed` section.
Notes in `Added` section are enough. But this is my option and maybe I am wrong?

tux-rampage replied:
I agree, in fact I was a bit unsure about it. I'd like to remove these entries for the same reason.
@Ocramius any objections?

weierophinney replied:
I agree with @webimpress. Instead, note that #32 and #34 provide the ability to log generator processing, and give a short example of how that can be enabled.

tux-rampage commented:
@Ocramius since you brought this up: Can you please approve?
I'd like to merge this by tomorrow, and update the changelog according to the suggestions from @webimpress, if there are no further objections.

weierophinney commented:
> edit: thank's github for mentioning internal ...

Surround annotation markings with code markers to avoid that: `@internal`. :smile:

weierophinney commented on CHANGELOG.md:
```diff
@@ -27,11 +27,14 @@ All notable changes to this project will be documented in this file, in reverse
 
-- Nothing.
+- [#33](https://github.com/zendframework/zend-di/pull/32) fixes discouraged use of
+  `Psr\Log\LoggerAware*` that was introduced with [#32](https://github.com/zendframework/zend-di/pull/32)
```
No need to add this here, as it was never previously released.

weierophinney commented on docs/book/codegen.md:
```diff
@@ -90,8 +90,8 @@ return [
-The `InjectorGenerator` implements the PSR-3 [`LoggerAwareInterface`](http://www.php-fig.org/psr/psr-3/#4-psrlogloggerawareinterface).
-So you can pass any PSR-3 logger to its `setLogger()` method.
+The `InjectorGenerator` allows to pass a [PSR-3 logger](http://www.php-fig.org/psr/psr-3/) as optional 
+fourth construction parameter.
```
Rephrase:

> The `InjectorGenerator` allows passing a PSR-3 logger instance via an optional, fourth constuctor parameter.

weierophinney commented on src/CodeGenerator/InjectorGenerator.php:
```diff
@@ -60,25 +61,33 @@ class InjectorGenerator implements LoggerAwareInterface
         ConfigInterface $config,
         DependencyResolverInterface $resolver,
-        ?string $namespace = null
+        ?string $namespace = null,
```
Just an FYI: you don't need to use a nullable type here when providing a default `null` value. (Something @webimpress demonstrated to me recently.)

weierophinney commented:
> Now I'm confused again. Since #34 is replacing #32, I'd just remove the refs to #32.

Just note whatever PRs led to the logger capabilities; I may be a bit fuzzy on which ones to use. :smile:

> > and give a short example of how that can be enabled.
> In the changelog? I'd just place a constructor example, would that be sufficient?

Perfect!

tux-rampage commented:
@weierophinney I incorporated the resquested changes. I also changed the title of this PR to reflect what was actually changed. Can you approve, please?

weierophinney commented on CHANGELOG.md:
```diff
@@ -6,8 +6,9 @@ All notable changes to this project will be documented in this file, in reverse
-- [#32](https://github.com/zendframework/zend-di/pull/32) adds the implementation of
-  `Psr\Log\LoggerAwareInterface` to `Zend\Di\CodeGenerator\InjectorGenerator`
+- [#34](https://github.com/zendframework/zend-di/pull/34) adds ability to pass
+  `Psr\Log\LoggerInterface` to `Zend\Di\CodeGenerator\InjectorGenerator`'s constructor 
```
Rephrase slightly:

> ... adds the ability to pass a `Psr\Log\LoggerInterface` instance to the constructor of `Zend\Di\CodeGenerator\InjectorGenerator`
